Protobuffer generation: support Int and UInt

nityanandaz opened the pull request:
## Problem that is solved

- Closes #140
- Create `IntegerWidthConfiguration` and `IntegerWidthCodingStrategy` to set a width preference for `Int` and `UInt`.

### Related PRs

- [x] Merge #131 first, because I restructured the tests there.
- [x] Merge #159, then configure `ProtobufferDecoder` and `ProtobufferEncoder`

### Testing

- [x] I will add tests that depend on the `Int.bitWidth`

### Reviewer Nudging

- [x] I am open for a better name for `isAmbiguousSupportedFixedWidthInteger`

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Handle many parameters

nityanandaz committed:
ProtobufferBuilderTests.testServiceParametersTwo fails

nityanandaz committed:
Make PropertyInfo editable (outside of Runtime)

nityanandaz committed:
Merge ProtobufferBuilderTests

nityanandaz committed:
Refactor ProtobufferBuilderTests.buildMessage

And add Node.collectValues,
delete ProtobufferBuilder.addService/addMessage

nityanandaz committed:
SwiftLint

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

# Conflicts:
#	Sources/Apodini/Semantic Model Builder/Model/TypeReflection/ParticularType.swift
#	Tests/ApodiniTests/ProtobufferBuilderTests/ProtobufferBuilderTests.swift

nityanandaz committed:
Supplement

nityanandaz committed:
Use mangledName(of:)

nityanandaz committed:
Remove differentiation

Trade client-side "convenience" for simplicity

nityanandaz committed:
Consider GRPCParameterOptions

PSchmiedmayer committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Refactor ProtobufferBuilder.analyze

nityanandaz committed:
Add comment

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Create isAmbiguousSupportedFixedWidthInteger

Move name strategy into Model+EnrichedInfo.swift

nityanandaz committed:
Pass tests

PSchmiedmayer commented on Sources/Apodini/Utilities/TypeInfo.swift:
```diff
@@ -50,3 +52,14 @@ func isSupportedScalarType(_ type: Any.Type) -> Bool {
         .contains(ObjectIdentifier(type))
 }
+
+func isAmbiguousSupportedFixedWidthInteger(_ type: Any.Type) -> Bool {
```
Maybe `isSupportedVariableWidthInteger` would be a more descriptive name?

PSchmiedmayer commented on Sources/Apodini/Semantic Model Builder/gRPC/ProtobufferBuilder/Extensions/Model+EnrichedInfo.swift:
```diff
@@ -35,20 +54,40 @@ extension ProtobufferMessage.Property {
+fileprivate extension TypeInfo {
+    func compatibleName() throws -> String {
+        if isSupportedScalarType(type) {
+            let infix = isAmbiguousSupportedFixedWidthInteger(type)
+                ? String(describing: Int.bitWidth)
+                : ""
+            let suffix = isSupportedScalarType(type)
+                ? ""
+                : "Message"
```
I would suggest adding if statements here to make the code a bit easier to read.

PSchmiedmayer reviewed (commented):
From a first glance those changes look good to me, I only had two small comments.

In addition to use `.bitWidth` to generate the correct protobuf description I would suggest to add some sort of configuration to the exporter (or Apodini in general) that would allow a developer to explicitly specify what type of `bitWidth` the developer would suspect to have when running the web service. 
If the `bitWidth` is different from what would be expected by the developer we should either fail at startup or at least log a logging error that the expected `bitWidth` of an `Int`/`UInt` differs from the expected platform as defined by the configuration.

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Supplement

nityanandaz committed:
Supplment 2

nityanandaz committed:
Refactor EndpointsTreeNode.collectAllEndpoints with `Node`

nityanandaz committed:
Rename collectAllEndpoints -> collectEndpoints

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/support-int-and-uint

# Conflicts:
#	Sources/Apodini/Utilities/TypeInfo.swift

nityanandaz committed:
Apply feedback

nityanandaz committed:
SwiftLint

PSchmiedmayer reviewed (approved):
Thanks for incorporating the feedback from the last review 🚀

nityanandaz committed:
Fix `Parameter` has closure property

https://github.com/Apodini/Apodini/pull/131#pullrequestreview-571524716

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

# Conflicts:
#	Sources/Apodini/Semantic Model Builder/Model/TypeReflection/Tree.swift
#	Sources/Apodini/Semantic Model Builder/gRPC/ProtobufferInterfaceExporter.swift

nityanandaz committed:
Move files, fix headers

nityanandaz committed:
Filter Handler Parameters from all other property wrappers, e.g., Environment.

nityanandaz committed:
Move .../gRPC/ProtobufferBuilder

nityanandaz committed:
Make ProtobufferInterfaceExporter a StaticInterfaceExporter

nityanandaz committed:
Remove private ProtobufferBuilderTests extension

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/support-int-and-uint

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

# Conflicts:
#	Sources/Apodini/Semantic Model Builder/Model/TypeReflection/EnrichedInfo.swift

nityanandaz committed:
Buff ProtobufferInterfaceExporter 💪

Nerf ProtobufferBuilder 🚶‍♂️

nityanandaz committed:
Merge branch 'protobuffer-generation/handle-many-parameters' into protobuffer-generation/support-int-and-uint

# Conflicts:
#	Sources/Apodini/Semantic Model Builder/Protobuffer/Model+EnrichedInfo.swift
#	Sources/Apodini/Semantic Model Builder/Protobuffer/TypeInfo+NameStrategy.swift

nityanandaz committed:
Supplement

nityanandaz committed:
Merge branch 'protobuffer-generation/support-int-and-uint' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Pass tests

Previously, scalar parameter messages were treated as composite message in the handler message

nityanandaz committed:
SwiftLint

nityanandaz committed:
Words

nityanandaz committed:
Create and pass testVariableWidthInteger

nityanandaz committed:
Supplement

nityanandaz committed:
Merge branch 'protobuffer-generation/handle-many-parameters' into protobuffer-generation/support-int-and-uint

nityanandaz committed:
Consistency

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Add tree documentation

nityanandaz committed:
Revise tree documentation

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/handle-many-parameters

nityanandaz committed:
Merge branch 'protobuffer-generation/handle-many-parameters' into protobuffer-generation/support-int-and-uint

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/support-int-and-uint

nityanandaz committed:
Create VariableWidthIntegerConfiguration

nityanandaz committed:
Use VariableWidthIntegerConfiguration

nityanandaz committed:
Create testVariableWidthIntegerConfiguration

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/support-int-and-uint

nityanandaz committed:
Parameterize ProtobufferInterfaceExporter.Builder

nityanandaz committed:
Expectation for testVariableWidthIntegerConfiguration

nityanandaz committed:
Pass new test

nityanandaz committed:
Adapt test; fix regression

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/support-int-and-uint

nityanandaz committed:
Fail larger VariableWidthIntegerConfiguration

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/support-int-and-uint

nityanandaz committed:
Create ProtobufferCoding.VariableWidthIntegerStrategy

nityanandaz committed:
Pass configuration/strategy to ProtobufferEncoder

nityanandaz committed:
Docs

nityanandaz committed:
Adapt InternalProtoEncodingContainer

nityanandaz committed:
Supplement

moritzschuell commented on Sources/Apodini/Configurations/VariableWidthIntegerConfiguration.swift:
```diff
@@ -0,0 +1,37 @@
+
+public struct VariableWidthIntegerConfiguration: Configuration {
+    // MARK: Nested Types
+    public enum Option: Int {
```
nit: Maybe a name more precise than `Option` would be better? Something like `IntWidthOption` 🤔

moritzschuell commented on Sources/Apodini/Semantic Model Builder/Protobuffer/ProtobufferInterfaceExporter.swift:
```diff
@@ -10,7 +10,9 @@ class ProtobufferInterfaceExporter: StaticInterfaceExporter {
 
-    internal enum Builder {}
+    enum Builder {
+        static var variableWidthIntegerPreference: Int?
```
Could we reuse `Option` (or `IntWidthOption` 😁) instead of the `Int` type? To enforce that only valid values (32 and 64) can be set?

moritzschuell commented on Sources/Apodini/Semantic Model Builder/gRPC/GRPCService/GRPCService.swift:
```diff
@@ -95,7 +99,17 @@ extension GRPCService {
-        let message = try ProtobufferEncoder().encode(AnyEncodable(value))
+        let encoder = ProtobufferEncoder()
+        if let configuration = app.storage[VariableWidthIntegerConfiguration.Key.self] {
+            let variableWidthIntegerStragegy: VariableWidthIntegerStrategy?
```
```suggestion
            let variableWidthIntegerStrategy: VariableWidthIntegerStrategy?
```

moritzschuell commented on Sources/ProtobufferCoding/Encoder/ProtoEncodingContainer.swift:
```diff
@@ -108,6 +108,16 @@ internal class InternalProtoEncodingContainer {
+            return
+        }
+        
         if MemoryLayout<Int>.size == 4 {
             try encodeInt32(Int32(value), tag: tag)
         } else if MemoryLayout<Int>.size == 8 {
```
Maybe instead of `encoder.variableWidthIntegerStrategy` being optional and keeping those `MemoryLayout.size` checks, we could make `encoder.variableWidthIntegerStrategy` non-optional and set it to the system's architecture-width by default in the constructor of the encoder?

nityanandaz replied:
I do that now: 40da4b12a554d66823168ba28484d7f3c668ce7f

moritzschuell reviewed (commented):
Great work @nityanandaz 🙌
Thanks for also adding the configuration functionality to the Proto en-/decoder!

I only left some small comments, once they are addressed I think this can be merged.

moritzschuell committed:
Add bit-width configuration also to Decoder

nityanandaz committed:
Remove indirection of VariableWidthIntegerConfiguration

nityanandaz committed:
Merge branch 'protobuffer-generation/support-int-and-uint' of github.com:Apodini/Apodini into protobuffer-generation/support-int-and-uint

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/support-int-and-uint

nityanandaz committed:
SwiftLint

moritzschuell committed:
Merge branch 'develop' into protobuffer-generation/support-int-and-uint

moritzschuell committed:
SwiftLint

nityanandaz committed:
Rename to VariableWidthIntegerCodingStrategy

nityanandaz committed:
Merge branch 'protobuffer-generation/support-int-and-uint' of github.com:Apodini/Apodini into protobuffer-generation/support-int-and-uint

# Conflicts:
#	Sources/ProtobufferCoding/Decoder/ProtoDecodingContainer.swift
#	Sources/ProtobufferCoding/Decoder/ProtobufferDecoder.swift

nityanandaz committed:
Update docs

nityanandaz committed:
SwiftLint

nityanandaz committed:
Update docs

lschlesinger commented on Sources/Apodini/Configurations/VariableWidthIntegerConfiguration.swift:
```diff
@@ -0,0 +1,38 @@
+    // MARK: Methods
+    public func configure(_ app: Application) {
+        guard rawValue <= Int.bitWidth else {
+            preconditionFailure(
```
This [helper for XCTAssertRuntimeFailure](https://github.com/Apodini/Apodini/blob/develop/Sources/XCTApodini/XCTAssertRuntimeFailure.swift) should help testing this precondition...

nityanandaz replied:
Thanks for pointing out the function. I can use it, but it sadly wouldn't be of any use in the near future, if I haven't missed something:

1. `XCTAssertRuntimeFailure` depends on `CwlPreconditionTesting`
2. `CwlPreconditionTesting` depends on `platforms: [.macOS]` in `Package.swift`
3. Macs will most likely never be 32-bit again.

Therefore, we cannot trigger the failure in a test case 😅

lschlesinger replied:
Haha 😆  I didn't really consider what is being checked here, my comment was only triggered by the uncovered precondition 🤦🏽‍♀️

nityanandaz replied:
The final names are `IntegerWidthConfiguration` and `IntegerWidthCodingStrategy` 👍

lschlesinger reviewed (approved):
I could not find any potential for code improvements. Just a hint for testing the precondition, if you weren't already aware of this. Regarding the coverage in general, I would guess it is not too hard to add tests cases (similar to `testVariableWidthIntegerConfiguration()`) with varying configurations in combination with `Int`, `UInt`, `[Int]`etc.? Maybe I am missing the difficulties in doing so 😄 .
 One comment regarding the variable naming, I think I would prefer `IntWidthConfiguration` instead of `VariableWidthIntegerConfiguration` this would then also be consistent with @moritzschuell 's [suggestion](https://github.com/Apodini/Apodini/pull/150/#discussion_r568047310) to rename `Option` to `IntWidthOption`, wouldn't it? I think it's easier to read as it is used very often and still captures the essence.

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/support-int-and-uint

nityanandaz committed:
Rename to IntegerWidthConfiguration

And Key to StorageKey

nityanandaz committed:
Rename to IntegerWidthCodingStrategy

nityanandaz committed:
Make InternalProtoEncoder.integerWidthCodingStrategy non-optional

nityanandaz committed:
Make ProtobufferInterfaceExporter.Builder a struct

A singleton does not work for the tests. They create race conditions

nityanandaz committed:
Create IntegerWidthConfiguration.default

nityanandaz committed:
Create IntegerWidthCodingStrategy.default

nityanandaz committed:
Codecov

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/support-int-and-uint

nityanandaz committed:
Create ProtobufferCoderTests to increase test coverage

nityanandaz committed:
Propagate strategy

nityanandaz committed:
Codecov

nityanandaz committed:
SonarCloud

nityanandaz committed:
SwiftLint

nityanandaz committed:
Codecov

moritzschuell reviewed (approved):
Thank you for implementing the suggested changes.
I have no more comments 😁

nityanandaz committed:
Merge branch 'develop' into protobuffer-generation/support-int-and-uint
